### DISCUSSION TITLE
TSA tries to handshake with long-gone workers

### DISCUSSION THREAD

**Discussion by vlfig-iov42** [concourse/concourse#7199] [Help & Support] [answered]
Hello wonderful people,

We're seeing in the web node logs repeated `tsa.connection.handshake-failed` messages for the same three long gone workers: `10.10.31.211`, `10.10.21.24` and `10.10.11.237` several times per second. These probably were alive during a recent (heavy) upgrade.

`fly workers` reports the current ones and only those. There's no other symptom other than stuffed logs but I wonder if you'd have an idea where to look for some stale data to delete?

/bows

```
Jun 22 10:10:05 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:05.692021034Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.31.211:53911","session":"4401"}}
Jun 22 10:10:05 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:05.692471320Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.31.211:1455","session":"4402"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.126626592Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.21.24:47132","session":"4403"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.137131126Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.11.237:21543","session":"4404"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.143138658Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.21.24:28542","session":"4405"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.232722590Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.11.237:27114","session":"4406"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.408492209Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.31.211:8217","session":"4407"}}
Jun 22 10:10:06 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:06.498630914Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.31.211:21914","session":"4408"}}
Jun 22 10:10:07 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:07.023127477Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.21.24:52961","session":"4409"}}
Jun 22 10:10:07 ip-10-10-31-199 concourse[11076]: {"timestamp":"2021-06-22T10:10:07.024470778Z","level":"info","source":"tsa","message":"tsa.connection.handshake-failed","data":{"error":"EOF","remote":"10.10.11.237:63749","session":"4410"}}
```

**Comment by taylorsilva** [comment, marked as answer]
Hello! `fly prune-worker` is the command you need. You can use it to tell the web node to forget about those workers.

Now for the **why**! The web nodes are not forgetting about these workers because they suddenly disappeared. If they were properly [retired](https://concourse-ci.org/concourse-worker.html#gracefully-removing-a-worker) then the web nodes would forget about them. When they suddenly disappear Concourse waits for them to come back. This ensures that things like temporary network outages don't mess up your cluster or kill inflight builds. We could probably make this smarter and have Concourse forget about workers after a few hours of no communication.

**Reply by vlfig-iov42** [reply]
Thanks Taylor,

To `fly prune-worker` I can only specify either a worker name or `--all-stalled`, but none of those failed handshakes in the logs are for a worker that currently exists (stalled or otherwise). They may have existed, but there's no trace of them anywhere.

`fly workers` only reports the current ones, the healthy, alive ones.

I've tried to search for those IP addresses in the database but couldn't find anything. I'll try grepping a `pgdump` just in case. Is there any other possible source of state where those IP addresses may have been left?

**Reply by taylorsilva** [reply]
Ohhh I see I had this backwards. So it seems you have someone trying to connect to the ssh server (TSA). Maybe a bad actor on your network? Or a worker installation your forgot about? I would look into what's running on the VM's owned by those IP's to see who's trying to establish a connection with the TSA on your web nodes.

**Reply by vlfig-iov42** [reply]
Oh, _I_ was seeing it backwards. I was thinking `web` was initiating connections. Not at all.

So that helped me figure it out: those were the internal LB's healthchecks _hammering_ TSA (:2222). I switched them to 8080 to get some peace and quiet.
